src/main/java/com/couchbase/lite/replicator/PullerInternal.java

```diff
@@ -456,10 +459,27 @@ private void queueDownloadedRevision(RevisionInternal rev) {
             }
         }
 
-        if (rev != null && rev.getBody() != null)
-            rev.getBody().compact();
+        final Body body = rev.getBody();
+        if (rev != null && body != null)
+            body.compact();
+
+        synchronized (downloadsToInsert) {
+            globalMemorySize+=body.size();
+        }
 
         downloadsToInsert.queueObject(rev);
+        if (tooManyQueuedObjects()) {
+            Log.d(TAG, "Flushing queue global memory size at " + globalMemorySize);
+            synchronized (downloadsToInsert) {
+                downloadsToInsert.flushAll(true);
```

Comment:
The solution is flush all queued task which includes data in the Batcher. I understand this solution is effective, but I don't have confidence if this is the right approach.